src/frontend/src/utils/validateDerivationOrigin.test.ts

```diff
@@ -0,0 +1,228 @@
+      fetchUrl: `https://totally-custom.com:8080/.well-known/ii-alternative-origins?canisterId=${TEST_CANISTER_ID}`,
+    },
+  ];
+
+  for (const { iiUrl, fetchUrl } of testCases) {
+    const fetchMock = setupMocks({
+      iiUrl,
+      response: Response.json({
+        alternativeOrigins: ["https://example.com"],
+      }),
+    });
+
+    const result = await validateDerivationOrigin({
+      requestOrigin: "https://example.com",
+      derivationOrigin: `https://some-url.com`, // different from requestOrigin so that we need to fetch the alternative origins
+      resolveCanisterId: () => Promise.resolve({ ok: TEST_CANISTER_ID }),
+    });
+
+    expect(result).toEqual({ result: "valid" });
+    expect(fetchMock).toHaveBeenLastCalledWith(fetchUrl, FETCH_OPTS);
+  }
+});
+
+test("should fetch alternative origins file using non-raw URL", async () => {
+  const fetchMock = setupMocks({
+    iiUrl: "https://identity.ic0.app",
+    response: Response.json({
+      alternativeOrigins: [`https://${TEST_CANISTER_ID}.raw.ic0.app`],
+    }),
+  });
+
+  const result = await validateDerivationOrigin({
+    requestOrigin: `https://${TEST_CANISTER_ID}.raw.ic0.app`,
+    derivationOrigin: `https://some-url.com`, // different from requestOrigin so that we need to fetch the alternative origins
+    resolveCanisterId: () => Promise.resolve({ ok: TEST_CANISTER_ID }),
+  });
+
+  expect(result).toEqual({ result: "valid" });
+  expect(fetchMock).toHaveBeenLastCalledWith(
+    `https://${TEST_CANISTER_ID}.icp0.io/.well-known/ii-alternative-origins`,
+    FETCH_OPTS
+  );
+});
+
+test("should not validate if canister id resolution fails", async () => {
+  const result = await validateDerivationOrigin({
+    requestOrigin: "https://example.com",
+    derivationOrigin: "https://derivation.com",
+    resolveCanisterId: () => Promise.resolve("not_found"),
+  });
+  expect(result.result).toBe("invalid");
+});
+
+test("should not validate if origin not allowed", async () => {
+  setupMocks({
+    iiUrl: "https://identity.ic0.app",
+    response: Response.json({
+      alternativeOrigins: ["https://not-example.com"],
+    }),
+  });
+
+  const result = await validateDerivationOrigin({
+    requestOrigin: "https://example.com",
+    derivationOrigin: `https://some-url.com`, // different from requestOrigin so that we need to fetch the alternative origins
+    resolveCanisterId: () => Promise.resolve({ ok: TEST_CANISTER_ID }),
+  });
+
+  expect(result.result).toBe("invalid");
+});
+
+test("should not validate if alternative origins file malformed", async () => {
+  setupMocks({
+    iiUrl: "https://identity.ic0.app",
+    response: Response.json({
+      notAlternativeOrigins: ["https://example.com"],
+    }),
+  });
+
+  const result = await validateDerivationOrigin({
+    requestOrigin: "https://example.com",
+    derivationOrigin: `https://some-url.com`, // different from requestOrigin so that we need to fetch the alternative origins
```

Comment:
Why is it using backticks instead of double quotes? Isn't prettier installed in the repo?

Reply:
We do have `prettier` but it doesn't care. Feel free to change the config. I'll fix the backticks here.